### PR TITLE
Add support for Razer Blade 14 (2022)

### LINKED ISSUE #1823: Support for Razer Blade 14'' with 3070Ti (2022)
### Device Info

* Name: Razer Blade 14'' with 3070Ti (2022)
* Product Number: RZ09-0427NEA3
* Model Number: RZ09-0427

### Device Information
```lsusb -d 1532: -v

Bus 001 Device 003: ID 1532:028c Razer USA, Ltd Razer Blade
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x028c 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Blade
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
```

```
ls -lah /dev/input/by-id/
total 0
drwxr-xr-x 2 root root 180 cze  1 13:20 .
drwxr-xr-x 4 root root 540 cze  1 13:20 ..
lrwxrwxrwx 1 root root   9 cze  1 13:20 usb-Azurewave_Integrated_RGB_Camera_01.00.00-event-if00 -> ../event4
lrwxrwxrwx 1 root root  10 cze  1 13:20 usb-Azurewave_Integrated_RGB_Camera_01.00.00-event-if02 -> ../event17
lrwxrwxrwx 1 root root   9 cze  1 13:20 usb-Razer_Razer_Blade-event-if01 -> ../event8
lrwxrwxrwx 1 root root   9 cze  1 13:20 usb-Razer_Razer_Blade-event-kbd -> ../event6
lrwxrwxrwx 1 root root   9 cze  1 13:20 usb-Razer_Razer_Blade-if01-event-kbd -> ../event7
lrwxrwxrwx 1 root root   9 cze  1 13:20 usb-Razer_Razer_Blade-if02-event-mouse -> ../event9
lrwxrwxrwx 1 root root   9 cze  1 13:20 usb-Razer_Razer_Blade-if02-mouse -> ../mouse2
```

### Packet Captures

WIP
<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
Screenshot (1)
Screenshot (2)
Screenshot (3)
Screenshot (4)
Screenshot (5)
Screenshot (6)
Screenshot (7)
Screenshot (8)
Screenshot (9)
Screenshot (10)
Screenshot (11)
Screenshot (12)
Screenshot (13)
Screenshot (14)
Screenshot (15)
Screenshot (16)
Screenshot (17)
Screenshot (18)

### PR DESCRIPTION
**Pull request by pramodhrachuri**
Made changes to v3.5.1 for fixing #1823 

Screenshot of working Polychromatic on Razer Blade 14 (2022):
image

My first-ever PR. Let me know if I made any mistakes in the procedure.

### CODE REVIEW

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -1011,6 +1012,10 @@ static ssize_t razer_attr_read_device_type(struct device *dev, struct device_att
         device_type = "Razer Blade 17 (2022)\n";
         break;
 
+    case USB_DEVICE_ID_RAZER_BLADE_14_2022:
+        device_type = "Razer Blade 14 (2022)\n";
```
Comment (turbos-and-tacos) [inline driver/razerkbd_driver.c:1016]: Do they do the "early" or "mid" or "late" designations for the 14s? If not, this is totally fine.

**Review by turbos-and-tacos — APPROVED** [review]
Did you give all the modes a quick test drive? The only reason I ask is because I saw the starlight effect acting very odd on a new device recently. Other than those two questions, LGTM!

**Comment by alkeryn** [comment]
also when doing pull requests, you shouldn't do merges but rebases ^^

**Comment by pramodhrachuri** [comment]
I'm currently traveling and don't have answers to the Qs. Will test all the modes after a day and update.
Edit: looks like 14s don't have any early/mid/late designations. They just have the years.

Sorry about the merge :/
How should we proceed then? Shall I make a second pull request and close this one?

**Comment by alkeryn** [comment]
@pramodhrachuri don't worry, you already did a great work ! :)
you could just reset hard to master, cherry pick the commits that fix the issue, then rebase onto master.
that way all your new commits will be on top of master with no merges in the history, then you can just force push ! :)

**Comment by z3ntu** [comment]
Don't worry about this, while it's nice if you could rebase it yourself, I can also easily fix this up during merging.

**Comment by alkeryn** [comment]
great !
btw @z3ntu i tested it on my razer blade 2022 3070ti, on nixos, i can confirm that the patch works.
only thing left to fix for that device is the microphone but that's out of topic haha.
though if you figured it out @pramodhrachuri i'd be glad to hear what you did.

also i could tell you how to fix the wifi if you didn't find, but i'm quite sure you did ! :)

**Comment by alkeryn** [comment]
wait i just found an issue with this patch.
all key light up but some symbols that did no longer do.
for example all special icons on the F* keys.
also on azerty keyboard the numbers do not light up, i'm thinking the special keys don't in qwerty keyboards.
also some specials characters don't light up.

all characters are supposed to be lit up on this device.

i'm thinking since older devices didn't have leds and this one now does, the issue arise, maybe it has more leds than previous devices.

it does suck because since i couldn't get the vm to work, now some keys simply are turned off.
well we'll have to find a fix for it, good work to begin though ! :)

**Comment by pramodhrachuri** [comment]
Interesting.
I didn't face any such problems. 
WhatsApp Image 2023-01-03 at 12 20 05 PM

https://user-images.githubusercontent.com/13718503/210408551-9a2c2050-e10c-4b74-b788-f619aed3823d.mp4

(This is the best I could do with the lighting. My surrounding is too bright.)

However, I just found something weird. When using the ripple effect, if I click the right ctrl key, the ripples start from the right arrow. I guess this has to do something with the key layout mapping. BTW, all other effects are getting applied including the starlight.


I haven't figured out the microphone yet :/  I now use my tab for meetings haha XD.
Yup, I have figured out the wifi. That was the 1st thing I tried to fix XD.

**Comment by alkeryn** [comment]
@pramodhrachuri now that is odd, maybe it has something to do with my keyboard being azerty ?

**Comment by alkeryn** [comment]
so i ended up doing a gpu passthrough to be able to run synapse in a vm and even through synapse i couldn't get the key to light back, so now i'm not sure they actually could light up on the french version, i'll have to look online.

Edit: Can confirm, it is just the french version being dumb, i could have swear i saw them light up but must be a wrong memory.

anyway, looks like all is well, i think this pr should be ready to merge ! :)

**Comment by sata1000000** [comment]
@pramodhrachuri Thats strange, your  windows key is different than mine.

PXL_20230103_233504376~2_nometa

**Comment by pramodhrachuri** [comment]
Wow. Mine looks like the Win 10 logo and yours is Win 11. I guess my laptop is just a little older than yours and Razer updated the logo in the newer batches. 
Hopefully, the logo print is the only change XD

**Comment by alkeryn** [comment]
@pramodhrachuri i bought mine like 3 weeks ago, a thing to mention was that usb4 came out of the box and i also have the win11 key.

i doubt there is much more change hardware wise, it is costly to change hardware and have to test things again, i don't think razer want to be in a situation where they changed some hardware in newer models only to have it cause issues later, unless there was already an issue and only a hardware fix could do it.

**Comment by z3ntu** [comment]
So functionally-wise everything is good from your side? Diff looks good.

**Comment by alkeryn** [comment]
@z3ntu yes, after a few weeks of testing everything looks good ! :)

**Comment by z3ntu** [comment]
Awesome, thanks everyone!

**Comment by mahdi65** [comment]
Any estimation of when razer blade 14 2023 will be added? Also, there is no sound in Ubuntu 20.04 5.15.0-75-generic

**Pull request merged**